src/core_modules/capture-core/components/WidgetEnrollment/EnrollmentDate/EnrollmentDate.container.js

```diff
@@ -0,0 +1,51 @@
+// @flow
+import React from 'react';
+import { useDataMutation } from '@dhis2/app-runtime';
+import { EnrollmentDateComponent } from './EnrollmentDate.component';
+import { convertValue as convertValueServerToClient } from '../../../converters/serverToClient';
+import { convertValue as convertValueClientToServer } from '../../../converters/clientToServer';
+import { dataElementTypes } from '../../../metaData';
+
+type Props = {
+    enrollmentDateLabel: string,
+    enrollment: any,
+    editEnabled: boolean,
+    displayAutoGeneratedEventWarning: boolean,
+    onUpdateDate?: (enrollmentDate: string) => void,
+    ...CssClasses,
+}
+
+const enrollmentUpdate = {
+    resource: 'tracker?async=false&importStrategy=UPDATE',
+    type: 'create',
+    data: enrollment => ({
+        enrollments: [enrollment],
+    }),
+};
+
+export const EnrollmentDate = ({
+    enrollmentDateLabel,
+    enrollment,
+    editEnabled,
+    displayAutoGeneratedEventWarning,
+    onUpdateDate,
+    classes,
+}: Props) => {
+    const [updateMutation] = useDataMutation(enrollmentUpdate);
+
+    const saveHandler = (selectedDate: string) => {
+        enrollment.enrolledAt = convertValueClientToServer(selectedDate, dataElementTypes.DATE);
+        updateMutation(enrollment);
+        onUpdateDate && onUpdateDate(enrollment.enrolledAt);
```

Comment:
Which part of the UI are you thinking of here? The date in the enrollment widget is rendered when the states `editMode` and `selectedDate` (in enrollmentDate.compontent) changes. Anyway, I think perhaps the change you had in mind got implemented in the error handling commit :slightly_smiling_face:

Reply:
Not exactly what I was looking for.

We should use optimistic UI and update the enrollment date instantly when the user changes it. Meaning; Before we get confirmation from the api, the enrollment date in the Widget UI should update and we should invoke the external onUpdate callback. If the request fails, we should revert the value to the original one. Also, in your saveHandler (EnrollmentDate.container) it is fine to invoke updateMutation directly, don't need to use an useEffect.

I think it makes most sense to update the enrollment object from useEnrollment (EnrollmentDate.container) instead of using a local component state.